### PR TITLE
Refactor magic methods, test massive assignment and validation

### PR DESCRIPTION
**Pull request by tom--**
I'm going to try the same refactoring on `__isset()` and `__unset()` so don't merge just now

### CODE REVIEW

**Comment by tom--** [comment]
the new magic methods all seem to work. the PR is ready for review!

File: tests/unit/DynamicActiveRecordTest.php [hunk 1]
```diff
@@ -42,130 +42,10 @@ protected function setUp()
...
-    {
-        $p = new Product();
-        $p->__set('4chan', 'sucks');
-    }
-
     public function testReadBinary()
     {
         $json = $this->db->createCommand(
-            'select column_json(dynamic_columns) from product where id=10'
+            'SELECT column_json(dynamic_columns) FROM product WHERE id=10'
```
Comment (tom--) [inline tests/unit/DynamicActiveRecordTest.php]: did i do this? if so, why?

File: tests/unit/DynamicActiveRecordTest.php [hunk 2]
```diff
@@ -394,7 +274,7 @@ public function testWriteRead($name, $value)
     {
         self::$resetFixture = false;
         $product = new Product();
-        /** @var string $product->$name */
+        /** @var string $product ->$name */
         $product->$name = $value;
```
Comment (tom--) [inline tests/unit/DynamicActiveRecordTest.php:278]: surely this is nonsense

File: tests/unit/DynamicActiveRecordTest.php (hunk 1 above)
Comment (djagya) [inline tests/unit/DynamicActiveRecordTest.php]: yes, it was so already when i started

File: composer.json [hunk 3]
```diff
@@ -27,7 +27,7 @@
     },
     "require-dev": {
         "yiisoft/yii2-dev": "~2.0",
-        "phpunit/phpunit": "~4.4"
+        "phpunit/phpunit": "~4.4",
```
Comment (djagya) [inline composer.json]: @tom-- it's invalid json now, remove comma

**Pull request merged**